File: 2023_07_31_functional_dependency_injection/fsharp_impl/FunctionalDependencyInjection/FunctionalDependencyInjection/RegisterVisitWithDeps.fs
module RegisterVisitWithDeps

open Domain
open Services


let getPatientById (patientRepository: IPatientRepository) patientId =
    match patientRepository.GetPatientById patientId with
    | None -> Error PatientNotFound
    | Some patient -> Ok(patient)


let addOrUpdatePatient
    (patientRepository: IPatientRepository)
    (logger: ILogger)
    getPatientById_
    (patient: PatientRegistrationInfoDto)
    =
    match patient with
    | Anonymous email -> Ok(None, email)
    | LoggedIn patientId ->
        getPatientById_ patientId |> Result.bind (fun patient -> Ok((Some patientId), patient.Email))
    | AnonymousWithCreatingNewAccount email ->
        let patientId = patientRepository.AddPatient { Id = ""; Name = email; Email = email }
        // let emailContent = $"New patient account has been created from '{email}'"
        // (env :> IEmailServiceP).EmailService.SendEmail(emailContent, email)
        Ok(Some patientId, email)
    | LoggedInWithUpdatingEmail (patientId, email) ->
        getPatientById_ patientId
        |> Result.bind (fun patient ->
            patientRepository.UpdatePatient { patient with Email = email }
            logger.Info $"Patient email has been changed from '{patient.Email}' to '{email}' "
            Ok((Some patientId), email))

let createPayment
    (externalPaymentService: IExternalPaymentService)
    (paymentRepository: IPaymentRepository)
    visitId
    price
    email
    =
    let externalPaymentId = externalPaymentService.CreatePayment(visitId, price, email)

    let payment =
        { Id = ""
          UserEmail = email
          Price = price
          VisitId = visitId
          ExternalPaymentId = externalPaymentId }

    paymentRepository.AddPayment payment |> ignore
    externalPaymentId


let registerVisit
    (scheduleRepository: IScheduleRepository)
    (emailService: IEmailService)
    addOrUpdatePatient_
    createPayment_
    (registration: RegisterVisitRequestDto)
    =
    match scheduleRepository.GetSlotById registration.SlotId with
    | None -> Error SlotNotFound
    | Some slot ->
        if not (Array.contains registration.ServiceId slot.ServiceIds) then
            Error SlotServiceNotFound
        else
            addOrUpdatePatient_ registration.Patient
            |> Result.bind (fun (patientIdO, email) ->
                let visit =
                    { Id = ""
                      SlotId = slot.Id
                      From = slot.From
                      To = slot.To
                      ServiceId = registration.ServiceId
                      UnitId = slot.UnitId
                      PatientId = patientIdO
                      PatientEmail = email }

                let visitId = scheduleRepository.AddVisit visit

                if slot.Price = 0m then
                    let emailContent = $"New visit created ... "
                    emailService.SendEmail(emailContent, email)

                    Ok { VisitId = visitId; ExternalPaymentId = None }
                else
                    let externalPaymentId = createPayment_ visitId slot.Price email

                    let emailContent = $"New visit created, please pay in 15 minutes here '{externalPaymentId}' "

                    emailService.SendEmail(emailContent, email)

                    Ok { VisitId = visitId; ExternalPaymentId = Some externalPaymentId })



let tests () =

    let (===) actual expected = if actual = expected then () else failwithf "assertion failed: %A <> %A" actual expected

    // let mutable appEnv = createEnvWithSlots ()
    let mutable registration = { SlotId = "1"; Patient = LoggedIn "1"; ServiceId = "1" }

    // let mutable appEnv = createEnvWithSlots ()

    let createAddOrUpdatePatient (appEnv: Env) =
        let getPatientById__ = getPatientById (appEnv.getService<IPatientRepository> ())

        let addOrUpdatePatient__ =
            addOrUpdatePatient
                (appEnv.getService<IPatientRepository> ())
                (appEnv.getService<ILogger> ())
                getPatientById__

        addOrUpdatePatient__

    let createRegisterVisit (appEnv: Env) =
        let addOrUpdatePatient__ = createAddOrUpdatePatient appEnv

        let createPayment__ =
            createPayment (appEnv.getService<IExternalPaymentService> ()) (appEnv.getService<IPaymentRepository> ())

        let registerVisit__ =
            registerVisit
                (appEnv.getService<IScheduleRepository> ())
                (appEnv.getService<IEmailService> ())
                addOrUpdatePatient__
                createPayment__

        registerVisit__


    // testing whole scenarios
    (createRegisterVisit (createEnvWithSlots ())) { registration with SlotId = "3" } === Error SlotNotFound

    (createRegisterVisit (createEnvWithSlots ())) { registration with ServiceId = "3" } === Error SlotServiceNotFound

    (createRegisterVisit (createEnvWithSlots ())) { registration with Patient = LoggedIn "2" } === Error PatientNotFound

    (createRegisterVisit (createEnvWithSlots ())) registration === Ok { VisitId = "1"; ExternalPaymentId = None }

    (createRegisterVisit (createEnvWithSlots ())) { registration with SlotId = "2" }
    === Ok { VisitId = "1"; ExternalPaymentId = Some "payment_1" }

    // appEnv <- createEnvWithSlots ()
    // registerVisit__ appEnv { registration with Patient = LoggedInWithUpdatingEmail("1", "najder.marcin@gmail.com") }
    // === Ok { VisitId = "1"; ExternalPaymentId = None }
    // (appEnv.getService<IEmailService> () :?> InMemoryEmailService).Store[0].Email === "najder.marcin@gmail.com"
    // appEnv.getService<IPatientRepository>().GetPatientById("1").Value.Email === "najder.marcin@gmail.com"


    // testing only subfunctions
    (createAddOrUpdatePatient (createEnvWithSlots ())) (Anonymous "marcin.najder@gmail.com")
    === Ok(None, "marcin.najder@gmail.com")

    (createAddOrUpdatePatient (createEnvWithSlots ())) (AnonymousWithCreatingNewAccount "marcin.najder@gmail.com")
    === Ok(Some "2", "marcin.najder@gmail.com")

    (createAddOrUpdatePatient (createEnvWithSlots ())) (LoggedIn "1") === Ok(Some "1", "marcin.najder@gmail.com")

    (createAddOrUpdatePatient (createEnvWithSlots ())) (LoggedIn "2") === Error PatientNotFound

    (createAddOrUpdatePatient (createEnvWithSlots ())) (LoggedInWithUpdatingEmail("1", "najder.marcin@gmail.com"))
    === Ok(Some "1", "najder.marcin@gmail.com")


    // passing dependencies as functions
    addOrUpdatePatient
        Unchecked.defaultof<_>
        Unchecked.defaultof<_>
        Unchecked.defaultof<_>
        (Anonymous "marcin.najder@gmail.com")
    === Ok(None, "marcin.najder@gmail.com")

    addOrUpdatePatient
        Unchecked.defaultof<_>
        Unchecked.defaultof<_>
        (fun id -> Ok({ Id = id; Name = ""; Email = "" }))
        (LoggedIn "1")
    === Ok(Some "1", "")

    addOrUpdatePatient Unchecked.defaultof<_> Unchecked.defaultof<_> (fun id -> Error PatientNotFound) (LoggedIn "1")
    === Error PatientNotFound

File: 2023_07_31_functional_dependency_injection/typescript_impl/src/registerVisitWithDeps.ts
import { Option, some, none, pipe, error, ok, Result, matchValue } from "powerfp";
import { Id, Patient, Payment, PatientRegistrationInfoDto, RegisterVisitRequestDto, RegisterVisitResponseDto, RegistrationError, Visit, Slot } from "./domain";
import { IEmailServiceP, IExternalPaymentServiceP, ILoggerP, IP, IPatientRepositoryP, IPaymentRepositoryP, IScheduleRepositoryP, createEnvWithSlots } from "./interfaces";
import * as assert from "assert";

type SkipFirst<T extends (...args: any) => any> = T extends (arg1: any, ...args: infer P) => infer R ? (...args: P) => R : any;

function getPatientById(env: IPatientRepositoryP, patientId: Id): Result<Patient, RegistrationError> {
    const patient = env.patientRepository.getPatientById(patientId);
    switch (patient.type) {
        case "some": return ok(patient.value);
        case "none": return error("PatientNotFound");
    }
}

function addOrUpdatePatient(
    env: { getPatientById: SkipFirst<typeof getPatientById> } & IPatientRepositoryP & ILoggerP,
    patient: PatientRegistrationInfoDto)
    : Result<[Option<Id>, string], RegistrationError> {

    switch (patient.type) {
        case "Anonymous": return ok([none, patient.email]);
        case "LoggedIn": return env.getPatientById(patient.patientId).bind(p => ok([some(p.id), p.email]));
        case "AnonymousWithCreatingNewAccount":
            const patientId = env.patientRepository.addPatient({ id: "", name: patient.email, email: patient.email });
            return ok([some(patientId), patient.email]);
        case "LoggedInWithUpdatingEmail":
            return env.getPatientById(patient.patientId)
                .bind(p => {
                    env.patientRepository.updatePatient({ ...p, email: patient.email });
                    env.logger.info(`Patient email has been changed from '${p.email}' to '${patient.email}' `);
                    return ok([some(p.id), p.email])
                });
    }
}




function createPayment(env: IExternalPaymentServiceP & IPaymentRepositoryP, visitId: string, price: number, email: string): string {
    const externalPaymentId = env.externalPaymentService.createPayment(visitId, price, email);
    let payment = { id: "", userEmail: email, price, visitId, externalPaymentId };
    env.paymentRepository.addPayment(payment)
    return externalPaymentId;
}


function registerVisit(
    env: { addOrUpdatePatient: SkipFirst<typeof addOrUpdatePatient>, createPayment: SkipFirst<typeof createPayment> } & IScheduleRepositoryP & IEmailServiceP,
    registration: RegisterVisitRequestDto): Result<RegisterVisitResponseDto, RegistrationError> {
    const slotO = env.scheduleRepository.getSlotById(registration.slotId);
    switch (slotO.type) {
        case "none": return error("SlotNotFound");
        case "some":
            if (slotO.value.serviceIds.indexOf(registration.serviceId) === -1) {
                return error("SlotServiceNotFound");
            } else {
                const slot = slotO.value;
                return env.addOrUpdatePatient(registration.patient)
                    .bind(([patientIdO, email]) => {
                        const visit: Visit = { id: "", slotId: slot.id, from: slot.from, to: slot.to, serviceId: registration.serviceId, unitId: slot.unitId, patientId: patientIdO, patientEmail: email }
                        const visitId = env.scheduleRepository.addVisit(visit);
                        if (slot.price === 0) {
                            const emailContent = "New visit created ... ";
                            env.emailService.sendEmail(emailContent, email)
                            return ok({ visitId, externalPaymentId: none }) as Result<RegisterVisitResponseDto, RegistrationError>;
                        } else {
                            const externalPaymentId = env.createPayment(visitId, slot.price, email);
                            const emailContent = `New visit created, please pay in 15 minutes here $'{externalPaymentId}' `;
                            env.emailService.sendEmail(emailContent, email);
                            return ok({ visitId, externalPaymentId: some(externalPaymentId) });
                        }
                    });
            }

    }
}


var appEnv = createEnvWithSlots();
let getPatientById_: SkipFirst<typeof getPatientById> = (...args) => getPatientById(appEnv, ...args);
let addOrUpdatePatient_: SkipFirst<typeof addOrUpdatePatient> = (...args) => addOrUpdatePatient({ ...appEnv, getPatientById: getPatientById_ }, ...args);
let createPayment_: SkipFirst<typeof createPayment> = (...args) => createPayment(appEnv, ...args);
let registerVisit_: SkipFirst<typeof registerVisit> = (...args) => registerVisit({ ...appEnv, createPayment: createPayment_, addOrUpdatePatient: addOrUpdatePatient_ }, ...args);



var registration: RegisterVisitRequestDto = { slotId: "1", patient: { type: 'LoggedIn', patientId: "1" }, serviceId: "1" };

type R = ReturnType<typeof registerVisit>;

assert(matchValue(registerVisit_({ ...registration, slotId: "3" }), error("SlotNotFound") as R));
assert(matchValue(registerVisit_({ ...registration, serviceId: "3" }), error("SlotServiceNotFound") as R));
assert(matchValue(registerVisit_({ ...registration, patient: { type: "LoggedIn", patientId: "2" } }), error("PatientNotFound") as R));

assert(matchValue(registerVisit_(registration), ok({ visitId: "1", externalPaymentId: none }) as R));


// testing only subfunctions
type R2 = ReturnType<typeof getPatientById>;

var noop: any = () => { };

assert(matchValue(
    getPatientById({ patientRepository: { getPatientById: id => none, addPatient: noop, updatePatient: noop } }, "1"),
    error("PatientNotFound") as R2
));

assert(matchValue(
    getPatientById({ patientRepository: { getPatientById: id => some({ id: "1", email: "", name: "" }), addPatient: noop, updatePatient: noop } }, "1"),
    ok({ id: "1", email: "", name: "" }) as R2
));


